javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
@shipilev many thanks for the great explanation!! :)

@paplorinc @jbgi then we should use a straight forward solution and add two optimizations

1) enable inlining by keeping the get() method small
2) add a local var that holds the volatile instance var within the synchronized block to reduce the volatile reads from 2 to only 1.

**Solution:**

```java
private volatile Supplier<? extends T> supplier;
private T value;

@Override
public T get() {
    return (supplier == null) ? value : computeValue();
}

private synchronized T computeValue() {
    final Supplier<? extends T> lazyValue = supplier;
    if (lazyValue != null) {
        value = lazyValue.get();
        supplier = null;
    }
    return value;
}
```

@paplorinc Could you please change the PR accordingly if the benchmarks still look good?

Thanks all! This will be an awesome Lazy implementation :)